docker/archive/multi_source.go

```diff
@@ -0,0 +1,62 @@
+package archive
+
+import (
+	"github.com/containers/image/v5/docker/internal/tarfile"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/types"
+	"github.com/pkg/errors"
+)
+
+// MultiImageSource manages a single Docker archive, allows listing its contents
+// and accessing individual images with less overhead than creating image references
+// individually (because the archive is, if necessary, copied or decompressed only once).
+type MultiImageSource struct {
+	path    string // The original, user-specified path; not the maintained temporary file, if any
+	archive *tarfile.ReadableArchive
+}
+
+// NewMultiImageSource returns a MultiImageSource for path.
+// The caller should call .Close() on the returned object.
+func NewMultiImageSource(sys *types.SystemContext, path string) (*MultiImageSource, error) {
+	archive, err := tarfile.NewReadableArchiveFromFile(sys, path)
+	if err != nil {
+		return nil, err
+	}
+	return &MultiImageSource{
+		path:    path,
+		archive: archive,
+	}, nil
+}
+
+// Close deletes temporary files associated with the MultiImageSource, if any.
+func (mis *MultiImageSource) Close() error {
+	return mis.archive.Close()
+}
+
+// List returns the a set of references for images in the MultiImageSource,
+// grouped by the image the references point to.
+// The references are valid only until the MultiImageSource is closed.
+func (mis *MultiImageSource) List() ([][]types.ImageReference, error) {
+	res := [][]types.ImageReference{}
+	for imageIndex, image := range mis.archive.Manifest {
+		refs := []types.ImageReference{}
+		for _, tag := range image.RepoTags {
```

Comment:
As mentioned in another comment, we need to account for manifests of untagged images where `RepoTags` is `nil`.